### ISSUE TITLE
Create from snapshot

### ISSUE THREAD

**Issue by femaref** [squat/terraform-provider-vultr#37]
Creation from snapshot works fine, e.g.:

```
data "vultr_os" "custom" {
  filter {
    name   = "family"
    values = ["snapshot"]
  }
}

data "vultr_snapshot" "master" {
  description_regex = "3_bootstrap_scripts"
}
```

however, once the servers are installed, `terraform plan` will return:

```
Terraform will perform the following actions:

  ~ vultr_instance.example[0]
      os_id: "159" => "164"

  ~ vultr_instance.example[1]
      os_id: "159" => "164"
```

It wants to change the os_id of the server from 159 (Custom) to 164 (Snapshot). It seems like the servers at vultrs side are actually listed as custom, not snapshot once they are fully installed.

I don't know if that's fixable or if a workaround in the tf file is needed, if even possible.

**Comment by squat** [comment]
Hi @femaref thanks for reporting this. I ran some tests to confirm this bug and it seems that servers that are restored from a snapshot are not always listed as having custom OS (159) but rather as having the OS of the original server that was snapshotted. To test this I:

1. brought up a server based on a snaphsot of a custom OS instance and Terraform reported its OS was 159 (custom) rather than 164 (snapshot). The Vultr console also showed the custom OS icon. This is the behavior you saw.
2. then created an Ubuntu server, snapshotted it, and created a server in Terraform from this snapshop. The OS was reported as 270 (Ubuntu) and sure enough, the Vultr console showed an Ubuntu icon for that server.

Clearly, Vultr stores metadata about the original server along with the snapshot. Unfortunately, they don't surface this information in the API [0]; the only snaphot fields returned are:
```json
    "5359435d28b9a": {
        "SNAPSHOTID": "5359435d28b9a",
        "date_created": "2014-04-18 12:40:40",
        "description": "Test snapshot",
        "size": "42949672960",
        "status": "complete"
    }
```

Is it possible the snapshot you restored was from a custom OS as well?

Right now, the best solution I can think of is to ignore OS changes for servers that were created from snapshots, i.e. if the `os_id` provided in configuration is 164. The way the Vultr API works now makes it difficult for any tool that reconciles state to work around this issue. If the responses for listing snaphots showed the OS ID of the snapshot then we could design a more elegant workaround that checks that the server's ID is either 164 or the OS ID of the snapshot. I am filing an issue with Vultr so they can track this problem. I'll update this thread with any information I gather.

[0] https://www.vultr.com/api/#snapshot_snapshot_list

**Comment by femaref** [comment]
yes, that is indeed the case. The snapshot restored was from a custom OS.

**Comment by squat** [comment]
@femaref I opened and ticket with the Vultr development team and discussed this issue and possible solutions with them. The ultimate problem is that this provider and the Vultr API were modeling the OS of an instance differently. This provider assumed that the OS of an instance was dependent only on the OS selected at creation/update. The API, on the other hand, models the the OS of an instance as being dependent on _two_ things: the OS selected at creation AND the snapshot selected at creation. This turns out to be the case as well for applications, i.e. if a chosen snapshot is of an application server, then once the instance boots, its application ID will be that of the original snapshotted app.

There are two ways to solve this in this provider:
1. the `os_id` field of the `vultr_instance` resource continues to be mandatory. When creating an instance based on a snapshot, rather than using the OS ID for snapshots, 164, the user must provide the OS ID of the OS that the snapshot was based on, e.g. 270 (Ubuntu). The provider does the necessary magic to use OS ID 164 during bring up and allow the final OS to be the one provided. This is a minimal public API change for the provider, but requires some extra work on the user's behalf and is not very intuitive.
2. the `os_id` field of the `vultr_instance` resource becomes optional and is computed by the provider. When creating an instance based on a snapshot, the user must not supply an `os_id`. Once the instance is up, the `os_id` field is computed and set to the OS ID of the machine that the snapshot was based on. This is a slightly larger public API change but requires the user to configure fewer options.

I went with option 2 (implemented in 441156ba4098e8dfb089f939c664ee94c22e6f74) because it simplifies the interface of the provider and allows the user to do less work. From now on, when booting a snapshot, only the `snapshot_id` must be configured. The `os_id` will be computed. The [examples show this new pattern](https://github.com/squat/terraform-provider-vultr/blob/master/examples/snapshot/example.tf). Please try booting a snapshot using the latest source from master and let me know if you still have any issues.

As an aside, the Vultr team graciously agreed to add "OSID" and "APPID" fields to the snapshot resources returned by the API. This will allow users to better introspect and filter snapshots.
